### PR TITLE
Decouple JSON serializability from serialization

### PR DESCRIPTION
**Pull request by znation**
This change accomplishes a few goals in order to help preserve backward
compatibility with naive JSON serialization (passing something like
`list(sf)` directly into `json.dumps`) and consumers using such
serialization.

* Separates data and schema serialization (so that data can be sent
  schemaless if desired, with schema attached, or with schema sent
  separately, to support a backward-compatibility path for existing
  APIs).
* Separates serializability (conversion between rich types and a set
  of values that are naively serializable) from serialization
  (conversion between naively serializable and strings).
* Moves responsibility for serialization itself out of this API (and
  into something like Python's `json` module).
* Removes rapidjson dependency since serialization is now handled at the
  Python layer.
* Fixes a mismatch in flexible_type type name between NoneType and
  undefined.

### CODE REVIEW

File: oss_src/unity/extensions/json/decoder.cpp [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+  for (const auto& kv : dict) {
+    if (kv.first == key) {
+      return kv.second;
+    }
+  }
+  std::stringstream msg;
+  msg << "Expected key \"";
+  msg << key;
+  msg << "\" was not present in dictionary input.";
+  throw msg.str();
```
Comment (ylow) [inline oss_src/unity/extensions/json/decoder.cpp]: Preferred is 
```
log_and_throw(...); 
```

it is included by
```
#include <logger/logger.hpp> 
```

File: oss_src/unity/extensions/json/decoder.cpp [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    case 0:
+      // flexible_type
+      return variant_get_value<flexible_type>(v);
+    default:
+      throw "Type mismatch: expected flexible_type as input to flex_from_variant.";
+  }
+}
+
+static flex_string _flex_string_from_variant(const variant_type& v) {
+  flexible_type f = _flexible_type_from_variant(v);
```
Comment (ylow) [inline oss_src/unity/extensions/json/decoder.cpp]: This function can probably be replaced with just a call to
```
variant_get_value<flex_string>(v)
```
It does raise an exception of something like "Expecting blah but we got a bleh" if it does not actually contain a string.

File: oss_src/unity/extensions/json/decoder.cpp [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    case flex_type_enum::DICT:
+      return _dict_from_serializable(data, schema);
+    case flex_type_enum::DATETIME:
+    case flex_type_enum::IMAGE:
+      throw "Unexpected input to _any_from_serializable: serializable flex_type does not include datetime or Image types.";
+  }
+}
+
+static schema_t schema_from_flex_dict(const flexible_type& f) {
+  flex_dict d = f.get<flex_dict>();
```
Comment (ylow) [inline oss_src/unity/extensions/json/decoder.cpp]: add an assert on d is flex_dict
```
ASSERT_EQUALS(f.get_type(), flex_type_enum::DICT)
```
its in
```
#include <logger/assertions.hpp>
```

**Comment by ylow** [comment]
looks good. some minor changes.

**Comment by znation** [comment]
Thanks for the review @ylow -- feedback addressed in 9941eb1ba391f22b3a76b16cb0da0ba9a6afe058

**Pull request merged**